datafusion/execution/src/config.rs

```diff
@@ -500,7 +500,7 @@ impl SessionConfig {
     /// ```
     ///
     /// [^1]: Compare that to [`ConfigOptions`] which only supports [`ScalarValue`] payloads.
-    pub fn with_extension<T>(mut self, ext: Arc<T>) -> Self
+    pub fn with_extension<T>(&mut self, ext: Arc<T>) -> &mut Self
```

Comment:
This is an API change and some people might like to be able to create a `SessionConfig` with a builder style

I think we should add a new function for updating an existing `SessionConfig` like this
```
    pub fn set_extension<T>(&mut self, ext: Arc<T>) {
...
}
```
(ideally also with an example)


cc @milenkovicm  as we also discussed various ways to make adding extensions easier

Reply:
we talked about the other extensions

```rust
 self.options_mut().extensions ...
```

which i always mix with this extensions 😀

I wonder if it make sense to register extension with internal mutability at the session creation and mutate that extension from your optimizer/planner? 🤔

Reply:
@MichaelScofield  can you change the PR to not change the public API (keep `with_extensions` and add `set_extensions`)?

Reply:
@alamb I've changed the PR according to your suggestions, PTAL.

Reply:
@milenkovicm we are using the `SessionState` a little differently in that it's not created everytime.